### ISSUE TITLE
add methods similar to collections.namedtuple?

### ISSUE THREAD

**Issue by AlJohri** [pcattori/icicle#11]
1) we should provide a similar method to create a new dict as a copy with some changes. basically the [collections.namedtuple._replace](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace) method. this is also similar to the `copy` method on another library https://github.com/slezica/python-frozendict/blob/master/frozendict/__init__.py#L34

2) there was an issue on that same repository about what to do when calling copy with no arguments. I didn't quite understand it though: https://github.com/slezica/python-frozendict/issues/6

3) we should provide a method to create a dict from an existing iterable. basically the [collections.namedtuple._make](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._make) method

**Comment by pcattori** [comment]
1): I like the idea of an method like `replace`, though I might prefer to call it `update` to make it more analogous to `dict`. Though the existing functionality might be ok:
```python
new_fd = FrozenDict(old_fd, **updates)
```

2) They are saying that if no modifications are being made you can return the *object itself*, rather than a copy of the object. This is desirable for performance reasons. Though I would argue that if you just want a copy with no modifications, the existing functionality is sufficient:
```python
fd_copy = FrozenDict(fd)
```

3): As for creating a `FrozenDict` from an existing iterable, I think the existing functionality is sufficient:
```python
fd = FrozenDict(iterable) # `iterable` is an iterable of length-2, key/value pair tuples 
```

---

To summarize, (1) is debatable, but I think (2) & (3) are already handled by the current API.

**Comment by AlJohri** [comment]
1): I'm not a fan of `update` because it traditionally returns `None` although I see where you're coming from. I would thing we should stub an `update` method and just raise a `NotImplementedError` in case someone is trying to use this as a drop in replacement. I would also argue that perhaps there should be some way to remove values from the dict which would require another method entirely. Since `FrozenDict({'a': 1, 'b': 2}, **{'b': None})` won't work.

2): Ah, I understand - this is for existing code that starts off a function with a `dict.copy` method. In which case, we should probably have a noop copy method so its a "drop-in replacement" for existing dicts? I'm not sure.

```
def my_legacy_code(d):
    d = d.copy()
    ...
```

3): Yes, I didn't realize you can do `dict(dict())`.

```
>>> iterable1 = (('a',1),('b',2),)
>>> iterable2 = {'c':3,'d':4}
>>> dict(iterable1)
{'b': 2, 'a': 1}
>>> dict(iterable2)
{'c': 3, 'd': 4}
```

**Comment by AlJohri** [comment]
Also it comes to mind we may want to add `__copy__` and `__deepcopy__`:

>In order for a class to define its own copy implementation, it can define special methods __copy__() and __deepcopy__(). The former is called to implement the shallow copy operation; no additional arguments are passed. The latter is called to implement the deep copy operation; it is passed one argument, the memo dictionary. If the __deepcopy__() implementation needs to make a deep copy of a component, it should call the deepcopy() function with the component as first argument and the memo dictionary as second argument.

https://docs.python.org/3/library/copy.html

**Comment by AlJohri** [comment]
http://stackoverflow.com/questions/15214404/how-can-i-copy-an-immutable-object-like-tuple-in-python

So I keep going back to `tuples` or `namedtuples` since they seem like the only truly immutable thing implemented in stdlib. So if the values of the frozen dict are all immutable, you can return `self` (the same id, memory reference) - but if the the values of the frozen dict are mutable, you have to actually perform the copy.

**Comment by pcattori** [comment]
Definitely think convenience methods for operating on a `FrozenDict` and receiving a new `FrozenDict` with the specified changes is a good idea 👍 (Similar to [Scala's "Operation in Class Map"](http://docs.scala-lang.org/overviews/collections/maps))

Unfortunately, `del` won't work as it cannot return a value (eg. new copy of a `FrozenDict`). But that's ok. We can just make a `delete` method there.

**Comment by AlJohri** [comment]
[werkzeug.datastructures.ImmutableDict](https://github.com/pallets/werkzeug/blob/master/werkzeug/datastructures.py#L1537)

This class implements both `copy` and `__copy__`. It returns a mutable dict for `copy` and a noop for `copy.copy`.
